---
title: "RNA-seq data cleaning"
subtitle: "Counts to voom"
author: "Kim Dill-McFarland, kadm@uw.edu"
date: "version `r format(Sys.time(), '%B %d, %Y')`"
output:
  html_document:
    toc: yes
    toc_depth: 3
    toc_float:
      collapsed: no
  pdf_document:
    toc: yes
    toc_depth: '3'
editor_options:
  chunk_output_type: console
urlcolor: blue
---

```{r include=FALSE}
knitr::opts_chunk$set(fig.height=2.5) 
#set working dir to project
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```

# 3. RNA-seq data cleaning

## Overview

Here, we take a raw RNA-seq counts table through quality assessment, filtering, and normalization. For more information on how counts were generated, see our [SEAsnake pipeline][SEAsnake]. These data are human, bulk, paired-end RNA-seq, but this pipeline can be applied to other organisms or single-read libraries. 

## Load packages

Load the following packages. For more information on installation, see the [setup instructions][lesson0].

```{r}
#Data manipulation and plotting
library(tidyverse)
#RNAseq data
library(edgeR)
library(limma)
library(RNAetc)

#Note we do not load biomaRt because it has conflicts with the tidyverse.
# We instead call its functions with biomaRt::

# Set random seed for reproduciblity
set.seed(651)
```

## Read in data

Using the tidyverse `read_` functions you saw earlier today, read in the counts and sample metadata files.

```{r}
count <- read_csv("0_data/raw_counts.csv")
meta <- read_csv("0_data/metadata.csv")
```

## Quality-filter data
### Filter poor-quality libraries

We assess library quality using metrics from [samtools `flagstat`][flagstat] and [`picard`][picard]. Here, we will only look at total sequences for brevity but you can find more quality measures of interest such as percent alignment and median CV coverage in our full [tutorial][count_to_voom].

Using `ggplot`, let's plot the total sequences per library.

```{r seqs}
ggplot(meta, aes(x = libID, y = total_seq)) +
  geom_col()
```

The above plot is somewhat difficult to read so we modify it with some `ggplot` features you've seen before and some that are new.
 
```{r seqs2}
ggplot(meta, 
       #Reorder x axis by y-axis value
       aes(x = reorder(libID, total_seq), 
           y = total_seq)) +
  geom_col() +
  #Add cutoff line
  geom_hline(yintercept = 1E6, color = "red") +
  #Set theme
  theme_classic() +
  #Relabel axes
  labs(x = "Library", y = "Total sequences") +
  #Rotate x-axis labels
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
```

Overall we see that all libraries have more the 1 million sequences. Based on this and other quality metrics not shown here, we determine that all libraries pass quality check. 

### Filter non-protein-coding genes

Standard differential expression analyses focus on protein-coding genes as these RNA products are the most likely to result in a measurable phenotype. We annotate the counts table genes to their biotypes as well as additional identifiers with `biomaRt`.

First, load the gene data base from `biomaRt`. Note that here we use `biomaRt::function( )` because we did not load this package with `library( )` earlier. The `::` operator tells R which package contains the function and prevents downstream issues in this pipeline due to `dplyr::select` vs `biomaRt::select`.

```{r ensembl}
#Get database
ensembl <- biomaRt::useEnsembl(biomart="ensembl",
                               dataset="hsapiens_gene_ensembl",
                                mirror = "www")

#Get gene key
key <- biomaRt::getBM(attributes=c("hgnc_symbol", "gene_biotype"),
                      mart=ensembl)
```

Also, you could extract more than just the biotypes from `biomaRt`. See all the possible data with `listAttributes`. 

```{r}
attrib <- biomaRt::listAttributes(ensembl)
head(attrib)
```
 
Next, filter the key to protein-coding genes present in the RNA-seq count data.

```{r}
key.filter <- key %>% 
  #Filter genes in count table
  filter(hgnc_symbol %in% count$hgnc_symbol) %>% 
  #Filter protein-coding
  filter(gene_biotype == "protein_coding")
```

And filter the count table to genes in the protein-coding key.

```{r}
count.pc <- count %>% 
  filter(hgnc_symbol %in% key.filter$hgnc_symbol)
```

This removes `r nrow(count)-nrow(count.pc)` genes from this data set and leaves `r nrow(count.pc)` genes for analysis.

### Filter PCA outliers

Sometimes one or more libraries will appear as outliers in PCA. We define this as any library more than 3 standard deviations away from the mean PC1 and/or PC2.

First, we must reformat the counts data to a log counts per million (cpm) matrix in order for PCA to work. We also transpose it so that genes are columns and libraries are rows. Without transposing, we'd get 1 dot per gene instead of 1 per library.

```{r}
count.for.pca <- count.pc %>% 
  #move HGNC gene symbol to rownames
  column_to_rownames("hgnc_symbol") %>% 
  #convert to log cpm
  cpm(log=TRUE) %>% 
  #transpose
  t()

count.for.pca[1:3,1:3]
```

Then, we calculate PCA on these transformed data.

```{r}
PCA <- prcomp(count.for.pca, scale.=TRUE, center=TRUE)
```

A basic `ggplot` of these PCA data looks like so, but it does not tell us anything about outliers.

```{r}
#Determine percent explained for each PC
summary(PCA)$importance
#create labels with percent explained for the PC
PC1.label <- paste("PC1", " (", summary(PCA)$importance[2,1]*100, "%)", sep="")
PC2.label <- paste("PC2", " (", summary(PCA)$importance[2,2]*100, "%)", sep="")

ggplot(as.data.frame(PCA$x),
       aes(x=PC1, y=PC2)) +
  geom_point(size=3) +
  labs(x=PC1.label, y=PC2.label)
```

We need to calculate the means and standard deviations of our PCs and create a new variable to color the points in PCA. First, let's get the summary metrics from the PCA data contained in `PCA$x`.

```{r}
PCA.summ <- as.data.frame(PCA$x) %>% 
  #move rownames into column of data frame
  rownames_to_column("libID") %>% 
  #Pivot to get all PCs in 1 column
  pivot_longer(-libID) %>% 
  #calculate mean and sd
  group_by(name) %>% 
  summarize(meanPC = mean(value),
            sdPC = sd(value))
PCA.summ

#Extract a row by the PC name
PC1.summ <- PCA.summ[PCA.summ$name == "PC1",]
PC2.summ <- PCA.summ[PCA.summ$name == "PC1",]

#create variable for outliers outside 3 sd from mean on either PC1 of PC2
PCA2 <- as.data.frame(PCA$x) %>% 
  mutate(outlier = ifelse(
    #Higher than +3 sd
    PC1 > PC1.summ$meanPC + 3*PC1.summ$sdPC | 
    PC2 > PC2.summ$meanPC + 3*PC2.summ$sdPC | 
    #Lower than -3 sd
    PC1 < PC1.summ$meanPC - 3*PC1.summ$sdPC |
    PC2 < PC2.summ$meanPC - 3*PC2.summ$sdPC, 
    #If any of the above are TRUE, make the variable equal to this
    "outlier",
    #Otherwise (else), use this
    "okay"))
```

Then our PCA plot reveals that there are no outliers. Note we added some additional `ggplot` customization to beautify this plot.

```{r}
ggplot(PCA2, aes(x=PC1, y=PC2)) +
  geom_point(aes(color=outlier), size=3) +
  labs(x=PC1.label, y=PC2.label) +
  #Set theme
  theme_classic() +
  #set xy ratio
  coord_fixed(ratio=1) +
  #change colors
  scale_color_manual(values = c("okay"="grey70",
                                "outlier"="red"))
```

We recommend that you initially remove dramatic outliers but leave those that are borderline or questionable. Then, you can re-assess outliers after gene filtering and normalization. You may find that some are no longer outliers after these steps. If they are, you can return to this step and remove them before repeating subsequent steps. 

### Additional steps

This workshop covers the minimum quality control and filtering needed for RNA-seq libraries. Checkout our [full tutorial](https://github.com/BIGslu/tutorials/blob/main/RNAseq/2.Hawn_RNAseq_counts.to.voom.pdf) for additional QC metrics, batch effects, custom plotting functions, and more!

### Create DGEList

At this stage, we've completed sample filtering and can collapse our count and meta data into a single list object. This allows us to shorten our long object names as well as works efficiently with the remaining cleaning steps.

We merge into a DGEList object, which is `edgeR` format. We must convert the counts data frame to a matrix for this process.

```{r}
count.pc.mat <- count.pc %>% 
  column_to_rownames("hgnc_symbol") %>% 
  as.matrix()

dat <- DGEList(counts=count.pc.mat, samples=meta, genes=key.filter)
```

This object is a list of data frames. It is an`S3` object in R, because it has 3 dimensions: `list[dataframe[column]]`. You can access each data frame in the list with `$` similar to how you've used it to access columns within a data frame.

```{r}
typeof(dat)
names(dat)

#One data frame in dat
dat$samples
#One column in one data frame in dat
dat$samples$libID
```

### Filter low abundance genes

Low abundance (small counts) and rare genes (many 0 counts) are removed from the data because they:

* are unlikely to be significantly differentially expressed
* greatly inflate multiple comparison correction 
* often do not meet linear modeling assumptions regarding mean variance trends (*e.g.* because of small N, they show lower variance than what is expected for their mean expression)

Our goal is to remove genes in the lower left of the mean-variance plot because counts (x) and variance (y) are low *e.g.* these genes break the red mean variance trend line.

Here we co-opt `limma`'s `voom` function to make out mean-variance plot for us.

```{r fig.width=4}
mv1 <- voom(dat, plot=TRUE)
```

We use our custom function in the package `RNAetc` to retain only genes that are at least `min.CPM` counts per million in at least `min.sample` number of samples OR in at least `min.pct` percent of samples. Here, we use 0.5 CPM in at least 3 samples.

```{r}
dat.abund <- filter_rare(dat, min.CPM = 0.5, min.sample = 3,
                         gene.var="hgnc_symbol")
```

Plotting the filtered data, we see the red trend line no longer curves down on the left. There is still a bit of the downward tail of dots but this filtering is sufficient.

```{r fig.width=4}
mv2 <- voom(dat.abund, plot=TRUE)
```

This filtering generally results in the removal of around 25% of genes. There is no exact cutoff for this filtering, and you should try several cutoffs to observe the effects. In general, we use minimum CPM from 0.1 - 1, minimum samples around 3 for small data sets, or minimum samples from 5 - 10% in larger data sets. It is important to keep the minimum sample cutoff larger than your smallest group of interest, otherwise you may filter genes specifically associated with one group. For example, if you have 10 libraries with 5 each of media and stimulated, your `min.sample` value should not be > 5 or else you will remove genes only expressed in one condition.

You may also wish to look for specific genes of interest and ensure they are not being filtered. For example, we will analyze interferon gammea (IFNG) later in this workshop so we check that this gene is present after filtering.

```{r}
"IFNG" %in% rownames(dat.abund$counts)
```

## Normalize data
### Trimmed mean of M (TMM)

RNA-seq counts are not independent within a library and not comparable across libraries. A library with 1 million sequences will have higher counts for most genes than one with 1 thousand sequences. We correct for this aspect of the data with the following normalization steps.

TMM defines a reference sample from your data set as the one with the most representative expression for the overall data set. Specifically, the reference sample is the one whose upper quartile is closest to the overall data set upper quartile. The upper quantile is the value (`x`) where 75% of genes have expression < `x`. Thus, the reference sample is the sample whose `x` is the closest to mean(`x`) across all samples.

All other samples are considered test samples. For each test sample, a scaling factor is calculated based on the weighted mean of log ratios of representative genes between the test and reference. These representative genes are a subset of the data set, removing the highest and lowest expressed genes as well as genes with the highest and lowest log ratios. The exact genes used as representative genes for scaling are dynamic and specific to each test sample.

The calculated scaling factors are then applied to the counts table automatically in the `voom` step.

```{r tmm}
dat.abund.norm <- calcNormFactors(dat.abund, method = "TMM")
```

### voom aka log2 counts per million (CPM)

We continue normalization by converting counts to CPM within each sample, thus accounting for differential sampling depth. We also perform log2 transformation, because RNA-seq data are heavily right-skewed and thus, violate assumptions of normality.

```{r warning=FALSE}
as.data.frame(dat.abund$counts) %>% 
  rownames_to_column() %>% 
  pivot_longer(-rowname) %>% 
ggplot() +
  geom_histogram(aes(x=value), bins = 100) +
  theme_classic() +
  labs(x = "count", y = "occurences") +
  lims(x=c(0,1000))
```

`voom` performs both of these steps! We use `voomWithQualityWeights` to additionally calculate sample specific quality weights that can be of use as co-variates in linear modeling.

```{r voom, fig.height=3.5}
#define model matrix
mm <- model.matrix(~ condition, data=dat.abund.norm$samples)
dat.abund.norm.voom <- voomWithQualityWeights(
                           dat.abund.norm,
                           design=mm,
                           plot=TRUE)
```

## Final data structure

Let's review the final data structure. We access each data frame within the `Elist` object using `$`. The normalized log2 CPM expression data are contained in `E`.

```{r}
dat.abund.norm.voom$E[1:3,1:7]
```

Library and donor metadata are in `targets`.

```{r}
dat.abund.norm.voom$targets[1:3,1:10]
```

Gene metadata are in `genes`.

```{r}
dat.abund.norm.voom$genes[1:3,]
```

Gene-level quality weights are in `weights`.

```{r}
dat.abund.norm.voom$weights[1:3,1:3]
```

## Explorative analysis

To get an initial sense of the cleaned data, we plot some variables of interest. Here, we see a clear Mtb infection signal and no relationship to sex.

```{r}
#transpose normalized expression data
voom.for.pca <- t(dat.abund.norm.voom$E)
#Calculate PCA
PCA3 <- prcomp(voom.for.pca, scale.=TRUE, center=TRUE)

#create labels with percent explained for the PC
PC1.label <- paste("PC1", " (", summary(PCA)$importance[2,1]*100, "%)", sep="")
PC2.label <- paste("PC2", " (", summary(PCA)$importance[2,2]*100, "%)", sep="")

#Merge PCA results with metadata
PCA.meta <- as.data.frame(PCA3$x) %>% 
  rownames_to_column("libID") %>% 
  full_join(meta)

#color by condition
ggplot(PCA.meta, aes(x=PC1, y=PC2)) +
  geom_point(aes(color=condition), size=3) +
  labs(x=PC1.label, y=PC2.label) +
  #Set theme
  theme_classic() +
  #set xy ratio
  coord_fixed(ratio=1)

#color by sex
ggplot(PCA.meta, aes(x=PC1, y=PC2)) +
  geom_point(aes(color=sex), size=3) +
  labs(x=PC1.label, y=PC2.label) +
  #Set theme
  theme_classic() +
  #set xy ratio
  coord_fixed(ratio=1)
```

## Save

Write final voom object as `RData`

```{r}
save(dat.abund.norm.voom, file = "0_data/dat_voom.RData")
```

We also save information on a single gene, IFNG, for use in our next section on linear modeling. Note that this combines several of the `tidyverse` functions you saw this morning as well as introduces some new ones.

```{r}
# Get 
IFNG <- as.data.frame(dat.abund.norm.voom$E) %>% 
  #Move rownames into the data frame as a column
  rownames_to_column("hgnc_symbol") %>% 
  #Filter gene of interest
  filter(hgnc_symbol == "IFNG") %>% 
  #Pivot to long format so libID are in a single column
  pivot_longer(-hgnc_symbol, names_to = "libID", values_to = "E") %>% 
  #Join with sample metadat
  full_join(dat.abund.norm.voom$targets, by = "libID")

#View resulting data frame
IFNG

#Save as csv
write_csv(IFNG, file = "0_data/IFNG.csv")
```

# *Navigation*

* [Workshop index][index]
* Previous lesson: [Intro to the tidyverse][lesson2]
* Next lesson: [Linear modeling][lesson4.1]

***

[index]: https://bigslu.github.io/2022_ASM_Microbe_RNAseq/index.html
[lesson0]: https://bigslu.github.io/2022_ASM_Microbe_RNAseq/0_welcome/0_install.html
[lesson1]: https://bigslu.github.io/2022_ASM_Microbe_RNAseq/1_introR/1_intro-R.html
[lesson2]: https://bigslu.github.io/2022_ASM_Microbe_RNAseq/2_tidyverse/2_tidyverse.html
[lesson3]: https://bigslu.github.io/2022_ASM_Microbe_RNAseq/3_RNAseq_cleaning/3_RNAseq_cleaning.html
[lesson4.1]: https://bigslu.github.io/2022_ASM_Microbe_RNAseq/4_linear_models/4.1_linear-models.html
[lesson4.2]: https://bigslu.github.io/2022_ASM_Microbe_RNAseq/4_linear_models/4.2_linear_model_rnaseq.html